clouddrift/wavelet.py

```diff
@@ -0,0 +1,534 @@
+"""
+This module provides wavelet functions.
+"""
+
+import numpy as np
+from typing import Optional, Tuple, Union
+import xarray as xr
+import warnings
+from scipy.special import gamma as _gamma, gammaln as _lgamma
+
+
+def wavetrans(
+    x: np.ndarray,
+    wave: np.ndarray,
+    norm: Optional[str] = "bandpass",
+    boundary: Optional[str] = "mirror",
+    time_axis: Optional[int] = -1,
+    freq_axis: Optional[int] = -2,
+    order_axis: Optional[int] = -3,
+) -> np.ndarray:
+    """
+    Continuous wavelet transform.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Real- or complex-valued signals
+    wave : np.ndarray
+        A suite of Morse wavelets as returned by function morsewave. The dimensions
+        of the suite of Morse wavelets are typically (f_order, freq_axis, time_axis).
+        The time axis of the wavelets must be the last one and matches the length of the time axis of x.
+        The normalization of the wavelets is assumed to be "bandpassed", if not use kwarg norm="energy".
+    boundary : str, optional
+        The boundary condition to be imposed at the edges of the time series.
+        Allowed values are "mirror", "zeros", and "periodic".
+        Default is "mirror".
+    order_axis : int, optional
+        Axis of wave for the order of the wavelets (default is first or 0)
+    freq_axis : int, optional
+        Axis of wave for the frequencies of the wavelet (default is second or 1)
+    time_axis : int, optional
+        Axis on which the time is defined for x (default is last, or -1). The time axis of the
+        wavelets must be last.
+
+    Returns
+    -------
+    wtx : np.ndarray
+        Time-domain wavelet transforms of input x. w shape will be ((series_orders), order, freq_axis, time_axis).
+
+    Examples
+    --------
+    To write.
+
+    Raises
+    ------
+    ValueError
+        If the time axis is outside of the valid range ([-1, N-1]).
+        If the shape of time axis is different for input and wavelet.
+        If ``boundary not in ["mirror", "zeros", "periodic"]``.
+
+    See Also
+    --------
+    :func:`morsewave`, `morsefreq`
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
+    # Positions and time arrays must have the same shape.
+    if x.shape[time_axis] != wave.shape[-1]:
+        raise ValueError("x and wave time axes must have the same length.")
+
+    wave_ = np.moveaxis(wave, [freq_axis, order_axis], [-2, -3])
```

Comment:
This `wave_` doesn't seem to be used, can it be removed?

Reply:
hmmm it is meant to be used at line 121 where it put into the padded wavelet which I initialized and also called `wave_` by mistake. Yet the tests passed...